all: implement EIP-1153 transient storage

tynes opened the pull request:
Implements `TSTORE` and `TLOAD` as specified by the following EIP:

https://eips.ethereum.org/EIPS/eip-1153
https://ethereum-magicians.org/t/eip-1153-transient-storage-opcodes/553

This PR includes test coverage of these opcodes.

We are interested in having this included as part of Shanghai, as talked about at the EIP panel at Devcon. There is general community support by application layer developers, see the Ethereum Magicians thread. There are 4 implementations so far, see the links below.

Includes an extensive test suite for `ethereum/tests` here https://github.com/snreynolds/tests/pull/1

Link to Shanghai planning issue https://github.com/ethereum/go-ethereum/issues/25820
Replaces https://github.com/ethereum/go-ethereum/pull/24463

For additional information, see https://www.eip1153.com/

Additional implementations:
- https://github.com/NethermindEth/nethermind/pull/4126
- https://github.com/ethereumjs/ethereumjs-monorepo/pull/1768
- https://github.com/hyperledger/besu/pull/4118

The EIP itself takes into account the maximum amount of memory that can be allocated if an entire block's worth of gas is used to `TSTORE` data, see https://eips.ethereum.org/EIPS/eip-1153#security-considerations.

About 9.15MB can be allocated in theory, plus the runtime overhead. cc @MariusVanDerWijden 
```
30M gas * 1 TSTORE / 100 gas * 32 bytes / 1 TSTORE * 1MB / 2^20 bytes ~= 9.15MB
```


Co-authored-by: Sara Reynolds <snreynolds2506@gmail.com>

holiman commented on core/state/journal.go:
```diff
@@ -213,6 +218,14 @@ func (ch storageChange) dirtied() *common.Address {
+	s.getStateObject(*ch.account).setTransientState(ch.key, ch.prevalue)
+}
+
+func (ch transientStorageChange) dirtied() *common.Address {
+	return ch.account
+}
```
We need to look into this: I think it is incorrect to report the account as dirtied by a transient op. 

The `dirtied` address is added to the journal `dirties`, which is later used to figure out which accounts have been modified -- in the sense that the trie representation of the account is changed. But a transient op does not modify the trie representation. 

Putting this note here as a todo to investigate a bit further.

holiman replied:
Ah, we do need to put it in `journal.dirties` after all, because otherwise we won't call `finalise` at the end of the tx, and if we don't do that, the transient storage doesn't get reset between transactions.

holiman replied:
I'm not so sure that a `Storage` on the `stateObject` is the best representation of a transient storage. The state-objects live longer than a single transaction. Having the transient on the objects themselves just forces us to iterate through them at the end of each tx, just to clean up. 

The transient storage(s) are essentially a transaction-global map of `address` --> `Storage`. After each transaction, we could just nuke it out and prepare a fresh one. 

This is how we did with the accesslist. In `statedb.go`:
```
	// Per-transaction access list
	accessList *accessList
```
It uses the 'normal' journal, but it is otherwise it's own separate entity, not tied to stateobjects.

tynes replied:
This makes sense, it feels like it sits at a similar level of abstraction as the access list since it is applied on a per transaction basis. Thank you for the feedback, will update the design to follow this approach

holiman replied:
I am of the opinion that `dirtied()` should return `nil`. Setting a transient storage value does not make an account `dirty`, in the trie-sense. 
I'm open to being proven wrong

tynes replied:
I agree with this, a dirty account implies that it was mutated in some way

holiman commented on core/vm/opcodes.go:
```diff
@@ -117,6 +117,8 @@ const (
 	GAS      OpCode = 0x5a
 	JUMPDEST OpCode = 0x5b
 	PUSH0    OpCode = 0x5f
+	TLOAD    OpCode = 0xb3
```
These should probably be in a latter range, this range is the `0x50 -range`. So somewhere further down

holiman reviewed (commented):
A few comments. Generally the implementation looks pretty clean. 

Also, I don't think there's any point updating the testdata -- not sure what commit you are updating it to, but it's not the latest. If it's some branch with 1153-tests, then that will have to be done later on once it's made it into `develop` on tests.

tynes commented:
Thank you for the review @holiman. I've addressed your feedback and moved the API away from the `stateObject`.

To run the `ethereum/tests`, I used the following setup:
- https://github.com/tynes/retesteth/tree/eip1153
- https://github.com/snreynolds/tests/tree/filler-tests

```bash
./retesteth/retesteth -t GeneralStateTests/stEIP1153 -- \
    --testpath <path to ethereum/tests> \
    --all --vmtrace --clients t8ntooleip --singlenet London
```

Note that I was having trouble with the `London+1153` syntax, so I just built `evm` with 1153 activated in London to run these tests for simplicity. Not sure why it was giving me an error

holiman commented on core/state/transient_storage.go:
```diff
@@ -0,0 +1,50 @@
+// Copyright 2020 The go-ethereum Authors
```
```suggestion
// Copyright 2022 The go-ethereum Authors
```

holiman commented on core/vm/instructions.go:
```diff
@@ -527,8 +527,25 @@ func opSstore(pc *uint64, interpreter *EVMInterpreter, scope *ScopeContext) ([]b
+	return nil, nil
+}
+
+func opTload(pc *uint64, interpreter *EVMInterpreter, scope *ScopeContext) ([]byte, error) {
+	loc := scope.Stack.peek()
+	hash := common.Hash(loc.Bytes32())
+	val := interpreter.evm.StateDB.GetTransientState(scope.Contract.Address(), hash)
+	loc.SetBytes(val.Bytes())
+	return nil, nil
+}
+
+func opTstore(pc *uint64, interpreter *EVMInterpreter, scope *ScopeContext) ([]byte, error) {
+	if interpreter.readOnly {
+		return nil, ErrWriteProtection
+	}
+	loc := scope.Stack.pop()
+	val := scope.Stack.pop()
+	interpreter.evm.StateDB.SetTransientState(scope.Contract.Address(), loc.Bytes32(), val.Bytes32())
```
We've not really been consistent, but most often, the new opcodes have gone into `eips.go` instead of `instructions.go`. I'm not really sure if it's needed. 

I think the reason has been to make the changes less "touching everything", and more easy to added/removed/combine with other changes. Not saying you _have_ to move it, just raising the point, not sure what others think cc @karalabe

holiman reviewed (approved):
LGTM, a minor nit or two

holiman commented on core/state/statedb.go:
```diff
@@ -452,6 +456,29 @@ func (s *StateDB) Suicide(addr common.Address) bool {
 	return true
 }
 
+func (s *StateDB) SetTransientState(addr common.Address, key, value common.Hash) {
```
Needs a method-doc

holiman commented on core/state/statedb.go:
```diff
@@ -452,6 +456,29 @@ func (s *StateDB) Suicide(addr common.Address) bool {
+	s.transientStorage.Set(addr, key, value)
+}
+
+func (s *StateDB) GetTransientState(addr common.Address, key common.Hash) common.Hash {
```
Needs a method-doc

holiman commented on core/state/transient_storage.go:
```diff
@@ -0,0 +1,50 @@
+	return make(transientStorage)
+}
+
+func (t transientStorage) Set(addr common.Address, key, value common.Hash) {
```
method-doc

holiman replied:
```suggestion
// Set sets the transient-storage `value` for `key` at the given `addr`.
func (t transientStorage) Set(addr common.Address, key, value common.Hash) {
```

holiman commented on core/state/transient_storage.go:
```diff
@@ -0,0 +1,50 @@
+	t[addr][key] = value
+}
+
+func (t transientStorage) Get(addr common.Address, key common.Hash) common.Hash {
```
method-doc

holiman replied:
```suggestion
// Get gets the transient storage for `key` at the given `addr`.
func (t transientStorage) Get(addr common.Address, key common.Hash) common.Hash {
```

holiman commented on core/state/transient_storage.go:
```diff
@@ -0,0 +1,50 @@
+	return val[key]
+}
+
+func (t transientStorage) Copy() transientStorage {
```
method-doc

holiman replied:
```suggestion
// Copy does a deep copy of the transientStorage
func (t transientStorage) Copy() transientStorage {
```

holiman reviewed (approved):
LGTM
